pkg/loadbalancing/dnslb/route53/impl.go

```diff
@@ -10,60 +10,77 @@ import (
 	"github.com/aws/aws-sdk-go/service/route53"
 )
 
-func newRecordReadWriter(hostedZoneId string,
+func newRecordReadWriter(awsSession *session.Session, hostedZoneId string,
 	logger log.DebugLogger) (*RecordReadWriter, error) {
 	if hostedZoneId == "" {
 		return nil, errors.New("no hosted zone ID specified")
 	}
-	awsSession, err := session.NewSession(&aws.Config{})
-	if err != nil {
-		return nil, err
-	}
-	if awsSession == nil {
-		return nil, errors.New("awsSession == nil")
-	}
 	return &RecordReadWriter{
 		awsService:   route53.New(awsSession),
 		hostedZoneId: aws.String(hostedZoneId),
 		logger:       logger,
 	}, nil
 }
 
-func (rrw *RecordReadWriter) readRecord(fqdn string) ([]string, error) {
+// Insert double quotes if missing.
+func insertQuotes(value string) string {
+	if value[0] == '"' && value[len(value)-1] == '"' {
+		return value
+	}
+	return "\"" + value + "\""
+}
+
+// Strip double quotes if present.
+func stripQuotes(value string) string {
+	if value[0] == '"' && value[len(value)-1] == '"' {
+		value = value[1 : len(value)-1]
+	}
+	return value
+}
+
+func (rrw *RecordReadWriter) readRecords(fqdn string, recType string) (
+	[]string, time.Duration, error) {
 	if fqdn[len(fqdn)-1] != '.' {
 		fqdn += "."
 	}
 	input := &route53.ListResourceRecordSetsInput{
 		HostedZoneId:    rrw.hostedZoneId,
 		StartRecordName: aws.String(fqdn),
-		StartRecordType: aws.String("A"),
+		StartRecordType: aws.String(recType),
 	}
 	output, err := rrw.awsService.ListResourceRecordSets(input)
 	if err != nil {
-		return nil, err
+		return nil, 0, err
 	}
-	var ips []string
+	var ttl time.Duration
+	var records []string
 	for _, recordSet := range output.ResourceRecordSets {
-		name := *recordSet.Name
-		if name[0] == '"' && name[len(name)-1] == '"' {
-			name = name[1 : len(name)-1]
-		}
-		if name != fqdn || *recordSet.Type != "A" {
+		name := stripQuotes(*recordSet.Name)
+		if name != fqdn || *recordSet.Type != recType {
 			continue
 		}
+		if _ttl := time.Duration(*recordSet.TTL) * time.Second; _ttl > ttl {
+			ttl = _ttl
+		}
```

Comment:
doesnt this become a ratchet always increasing?

Reply:
It's just returning the largest TTL from all the record sets. It's not accumulating them,

Reply:
Route53 sometimes splits multiple records into different record sets, even if they were written in a single record set.
Since the method is returning a single TTL (how long records may stay in caches), the largest TTL of the record sets is what we want.